input-output-hk/cardano-sl
# [CSL-2247] Extend the V1 API to report restoration progress

## Summary

This PR (which is being opened against `feature/CBR-90` is virtually a feature-complete incarnation of CBR-90. *The only thing missing are DB migrations, which will be delivered with CSL-2429. This doesn't hold up the Daedalus team from start integrating the feature, as the V1 API provided here won't change*.

## Description

This PR builds upon CSL-2246 (for which the reader is encouraged to defer review up until a PR for `feature/CBR-90` won't land, to get the full picture). In a nutshell we extend the `WalletInfo` section of the `WalletStorage` to keep around some `SyncStatistics` which can be used by the V1 API to compute interesting numbers about a wallet `syncState`, for example:

- `SyncState` (Syncing? Restoring? etc)
- `SyncPercentage` (0 .. 100%)
- `EstimatedCompletionTime` (in milliseconds)
- `SyncThroughput` (in blocks/s)

Notably, `SyncProgress` has been renamed to `SyncPercentage` and `SyncProgress` is now a richer type.

## Screenshots

<img width="1915" alt="screen shot 2018-03-29 at 15 43 44" src="https://user-images.githubusercontent.com/29383371/38096454-3545f964-3373-11e8-9476-1d04206d6d26.png">

## QA Steps

Provided as part of `feature/CBR-90`, but for @darko-mijic and @nikolaglumac the best course of action is to attach the `cardano-node` to the staging cluster, like this:

```
stack build cardano-sl-wallet-new
stack exec cardano-node -- --topology=docs/network/example-topologies/mainnet-staging.yaml --configuration-key mainnet_dryrun_full --db-path ../staging-db
```

And finally play with the wallet. It's recommended to import a staging DB so that the restoration is realistic. **Note that transactions cannot be sent during a restoration (due to known limitations we discussed) so doing so will incur in unexpected behaviour.**
